Add TypeScript

spessasus opened the pull request:
This pull request adds TypeScript to spessasynth_core.

Fixes #8.

Fixes #10 

Fixes #12.

It also:
- Introduces a lot of breaking changes.
- improves the API consistency and makes sure that an InstrumentZone or PresetZone always has an element.
- Documents the breaking changes.
- Adds full API support for bank LSB.
- Adds tseslint and prettier, ensuring consistent code style.

@sylvia-leaf what do you think? Let me know when I can merge it.


To do:

- [x] - make sure it compiles
- [x] - enforce eslint rules and prettier for build
- [x] - make sure it works at least with spessafont
- [x] - document breaking changes
- [x] - rework sequencer to have an event handler like synthesizer
- [x] - update documentation     
- [x] - widen bank API to add LSB
- [x] - rework MIDIBuilder to properly handle format 1 conductor track & prevent adding more tracks to format 0

spessasus committed:
initial typescript

spessasus committed:
initial typescripting
doesn't build yet
addresses #8

spessasus committed:
further typescripting
sound banks + midi handlers completed
still doesn't build
addresses #8

spessasus committed:
Finish TypeScript
With a few API changes!

spessasus committed:
Finish TypeScript
With a few API changes!

sylvia-leaf commented:
@spessasus What needs to be done on my side to ensure that my code is compatible with the Typescript?

spessasus committed:
A few more API tweaks since it's already breaking changes

spessasus commented:
@sylvia-leaf 
What's nice about TS is that it works with all compatible editors, like VSCode (which is what I'm assuming you're using). JSDoc only worked consistently in webstorm according to my testing, so now you'll get full type support, which will help with understanding the code :-)

Regarding the changes:
Generally the soundfont loading code is the same, although there are a few differences:
- classes like `Preset` or `Instrument` are renamed to `SoundFontPreset` and `SoundFontInstrument` respectively. I'm not sure why I haven't done this earlier, as they are only for reading SF2s, not DLS or generic. Generic sound bank classes are still prefixed with `Basic`.
- `presetName`, `instrumentName` and `sampleName` have all been renamed to just `name`. I wasn't even thinking of making this a library when I first chose those names 2 years ago :-)
- similarly, `instrumentZones` and `presetZones` are just `zones` now.
- a few changes to `BasicSample` names, mostly removing `sample`: `samplePitchCorrection -> pitchCorrection`.
- And one change in the SF loading logic, though I don't think you've touched it: the instrument and preset zones now _must_ have their samples and instruments set on init. This changed the SF2 loading code slightly as we now search the generator list for `sampleID` and `instrument` before we add them to the zone.

I'd also like you to know that this will be released as 3.28, as I'm moving spessasynth_lib (which is very needed because of [this mess](https://github.com/spessasus/spessasynth_lib/blob/acc696ea5a0c2b4e5cd369bb2f22d35e35f4949f/src/synthetizer/worklet_message.js#L55-L108)) and SpessaSynth to typescript as well. I'll also need to tweak SpessaFont to work with the new API, but since it's already TS, I'm pretty sure that you will be able to just merge the changes. So there's no need to rush :-)

So, here's what I think you should do:
- Compare your fork with the current master (JS) and see the channges.
- Copy the `typescript` branch to your fork.
- `npm install`
- Apply the changes you've made to typescript.
- `npm run build` to test if all TS files are still valid. Examples have been adjusted so they work as before, but now you need to use `tsx` rather than `node`.
- Once that's done, you can do the same to lib and web app.


Another thing if you need it:
The best way to test typescripted core is to run `npm run build`, then `npm pack`. It will create a `.tgz` file (which is what gets uploaded to NPM). Then in lib, font or web app, `npm uninstall spessasynth_core`, then you can manually add:
```json
"dependencies" : {
    "spessasynth_core": "/path/to/your/file.tgz"
}
```
Then `npm install` and the typescript version will work like it was published.

Let me know if this helps!

spessasus committed:
Fix types for package.json

spessasus committed:
Soundbank instead of soundfont

spessasus committed:
fix MIDI CC gaps

spessasus committed:
Fix enum types

spessasus committed:
small tweak

spessasus committed:
event types camelCase

spessasus committed:
spessafont working

sylvia-leaf commented:
@spessasus The plan is to complete the UTF-8 support in my fork of spessasynth, and then start work on merging typescript.

spessasus committed:
API tweaks

spessasus committed:
Merge branch 'master' into typescript

# Conflicts:
#	.idea/codeStyles/codeStyleConfig.xml
#	.idea/inspectionProfiles/Project_Default.xml
#	README.md
#	package.json
#	src/synthetizer/audio_engine/engine_methods/tuning_control/set_modulation_depth.js

spessasus committed:
fix imports

spessasus committed:
Create a Migration Guide
And fix some implementation errors

spessasus committed:
update docs for event types

spessasus committed:
small tweaks

spessasus committed:
stricter eslint + fixes

spessasus committed:
eslint stylistic

spessasus committed:
better sequencer behavior

spessasus committed:
loop

spessasus committed:
fix sequencer

spessasus committed:
docs

spessasus committed:
readme

spessasus committed:
a few fixes and tweaks

spessasus committed:
Merge branch 'master' into typescript

# Conflicts:
#	docs/index.md
#	mkdocs.yml

spessasus committed:
it's 4.0

spessasus committed:
small fix

spessasus committed:
Sequencer and MIDI refactors

spessasus committed:
More tweaks

spessasus committed:
Better api and "sound bank"

spessasus committed:
small tweak!

spessasus committed:
Remove reloadManager!

spessasus committed:
Merge branch 'master' into typescript

spessasus committed:
fix types

spessasus committed:
better BasicMIDI

spessasus committed:
Capitalized comments

spessasus committed:
noinspection

spessasus committed:
revise binary functions

spessasus committed:
fix the DLS parser

spessasus committed:
small tweaks

spessasus committed:
remove channel configuration

spessasus committed:
fix no loops with loop on release
fixes https://github.com/spessasus/SpessaSynth/issues/174

spessasus committed:
make sure that all write methods return arraybuffers instead of uint8arrays
and add more breaking changes to the API :-)

spessasus committed:
small fixes

spessasus committed:
unrecognized is info now

spessasus committed:
tweaks to MIDI metadata

spessasus committed:
tweak

spessasus committed:
more tweaks!

spessasus committed:
better pitch bend

spessasus committed:
minor tweaks once more

spessasus committed:
tweak

spessasus committed:
tweak!

spessasus committed:
better write wav

spessasus committed:
More JSDoc and remove unneeded param

spessasus committed:
Minor tweaks :-)

spessasus committed:
midiData can be undefined!

spessasus committed:
soundBankInfo overhaul + Full RIFF FourCC typing!

spessasus committed:
sf2 write fix!

spessasus committed:
Date fix

spessasus committed:
better RMI defaults

spessasus committed:
better RMI defaults

spessasus committed:
remove compressionQuality from type

spessasus committed:
small safeguard

spessasus committed:
tweak

spessasus committed:
consistent pitch wheel

spessasus committed:
consistent pitch wheel & other tweaks

spessasus committed:
tiny change

spessasus committed:
remove name and better metadata

spessasus committed:
sort lyrics
fixes #10

spessasus committed:
Merge branch 'master' into typescript

spessasus committed:
enforce correct MIDI builder

spessasus committed:
stop all channel number

spessasus committed:
clone metadata from tracks too

spessasus committed:
rework the RMIDI INFO API

spessasus committed:
remove end trimming

spessasus committed:
better date parser

spessasus committed:
no globalization allowed

spessasus committed:
write default modulators only when needed

spessasus committed:
fix version write

spessasus committed:
initial lsb

spessasus committed:
fix invalid DLS articulator transform

spessasus committed:
Merge branch 'typescript' into typescript-lsb

spessasus committed:
LSB selection works

spessasus committed:
it builds now

spessasus committed:
fix preset shadowing

spessasus committed:
remove preset shadowing

spessasus committed:
fix DLS drums

spessasus committed:
harden patch selection system

spessasus committed:
harden patch selection system again

spessasus committed:
improve bank offset && fix SF2 write

spessasus committed:
patch up midi editor

spessasus committed:
namespace for bank select hacks

spessasus committed:
small tweak

spessasus committed:
small tweaks

spessasus committed:
explain bank LSB

spessasus committed:
Merge pull request #15 from spessasus/typescript-lsb

Add LSB to typescript

spessasus committed:
add LSB support for SFe read

addresses https://github.com/SFe-Team-was-taken/SFeReferenceImplementation_Core/issues/3

spessasus committed:
structure rework

spessasus committed:
initial DLS rework
does not build yet

spessasus committed:
partial SF conversion

spessasus committed:
a few more tweaks

spessasus committed:
important fixes

spessasus committed:
modulator source rework!
I know it's the DLS branch, but I wanted to!

spessasus committed:
Fix keyNumToPitch correction

spessasus committed:
Fix various DLS quirks

spessasus committed:
initial SF2 to DLS

spessasus committed:
polish up the DLS converter

spessasus committed:
move the flattening into the preset

spessasus committed:
small refactors

spessasus committed:
fix SF2 write

spessasus committed:
fix DLS control transform getting written

spessasus committed:
fix a broken fix!

spessasus committed:
reintroduce globalization

spessasus committed:
add instrument aliasing support
fixes https://github.com/spessasus/spessasynth_core/issues/14

spessasus committed:
types fix

spessasus committed:
quick DLS copy fix

spessasus committed:
validate DLS sound banks

spessasus committed:
Merge pull request #16 from spessasus/dls-rework

DLS Rework

spessasus committed:
update documentation

spessasus committed:
further documentation

spessasus committed:
rework the SF2 writing system

spessasus committed:
sound bank docs

spessasus committed:
finish docs!

spessasus committed:
remove some pointless properties
